### PR TITLE
Return a Result Monad when loading the config

### PR DESCRIPTION
**Pull request by screendriver**

### CODE REVIEW

File: test/unit/config.test.ts [hunk 1]
```diff
@@ -31,50 +32,91 @@ function createExplorer(overrides: Partial<ReturnType<typeof cosmiconfig>> = {})
...
+  test('returns an Result Err when no config exists', async function () {
     const explorer = createExplorer({
-      load: sinon.fake.rejects(new Error()),
+      load: sinon.fake.rejects(new Error('Not found')),
     });
-    const actual = await loadConfig('./not-found.js', explorer);
-    const expected = undefined;
+    const config = await loadConfig('./not-found.js', explorer);
+    if (config.isOk()) {
+      assert.fail('loadConfig() did not returned an expected error');
```
Comment (lo1tuma) [inline test/unit/config.test.ts:42]: You could avoid the `if` statement by using `assert.strictEqual(config.isOk(), false, 'loadConfig() did not returned an expected error')`.

File: test/unit/config.test.ts [hunk 2]
```diff
@@ -31,50 +32,91 @@ function createExplorer(overrides: Partial<ReturnType<typeof cosmiconfig>> = {})
...
+    }
+    const actual = config.error;
+    const expected = 'Not found';
     assert.strictEqual(actual, expected);
   });
 
-  test('loadConfig() returns undefined when config is empty', async function () {
+  test('returns an Result Err when loading does not return an error object', async function () {
```
Comment (lo1tuma) [inline test/unit/config.test.ts:49]:
```diff
-returns an Result Err when loading does not return an error object
+returns a Result Err when loading does not return an error object
```

File: test/unit/config.test.ts (hunk 1 above)
Comment (screendriver) [reply test/unit/config.test.ts:42]: Unfortunately not. TypeScript's control flow analysis is not smart enough here and only works with an `if` statement 😕 So with this "inline" assertion the compiler still thinks that property 'error' does not exist.

File: test/unit/config.test.ts (hunk 2 above)
Comment (screendriver) [reply test/unit/config.test.ts]: Fixed in 2da9cd8d2966fe3ccee8f5e7b798e69e49b2f87b

File: test/unit/config.test.ts (hunk 1 above)
Comment (lo1tuma) [reply test/unit/config.test.ts:42]: Hopefully this will be fixed in the next version of typescript.

**Review by lo1tuma — APPROVED** [review]
LGTM

**Pull request merged**